PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -204,80 +241,178 @@ public void AttachToRunspace(uint pid)
+                    if (preScenario == postScenario)
                     {
-                        return true;
+                        // very unlikely for this to happen, but we should make sure to handle the case anyway
+                        ServiceCommon.Log("Failed to detach from running process");
+                        return false;
                     }
                 }
+
+                // rehook event handlers
+                AddEventHandlers();
             }
-            catch (Exception ex)
+
+            return true;
+        }
+
+        /// <summary>
+        /// Finds all processes on a remote computer which have loaded the powershell.exe module
+        /// </summary>
+        /// <param name="remoteName">Name of the remote machine</param>
+        /// <returns></returns>
+        public List<KeyValuePair<uint, string>> EnumerateRemoteProcesses(string remoteName)
+        {
+            List<KeyValuePair<uint, string>> validProcesses = new List<KeyValuePair<uint, string>>();
+
+            // Retrieve callback context so credentials window can display
+            if (_callback == null)
             {
-                ServiceCommon.Log(string.Format("{0} , cannot examine modules of process; id: {1}", ex.Message, pid));
+                _callback = OperationContext.Current.GetCallbackChannel<IDebugEngineCallback>();
             }
-            return false;
+
+            using (_currentPowerShell = PowerShell.Create())
+            {
+                try
+                {
+                    // Initiate remote session with the remote machine
+                    _currentPowerShell.Runspace = _runspace;
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.EnterRemoteSessionDefaultCommand, remoteName));
+                    _currentPowerShell.Invoke();
+
+                    if (GetDebugScenario() == DebugScenario.Local)
+                    {
+                        return null;
+                    }
+
+                    // Run script on remote machine to grab all attachable processes
+                    _currentPowerShell.Commands.Clear();
+                    _currentPowerShell.AddScript(DebugEngineConstants.EnumerateRemoteProcessesScript);
+                    _currentPowerShell.Runspace = _runspace;
+                    Collection<PSObject> result = _currentPowerShell.Invoke();
+
+                    // Add each process' name and pid to the list to be returned
+                    for (int i = 0; i < result.Count; i += 2)
+                    {
+                        dynamic pid = result.ElementAt(i);
+                        dynamic name = result.ElementAt(i + 1);
+                        validProcesses.Add(new KeyValuePair<uint, string>(uint.Parse(pid.ToString()), name.ToString()));
+                    }
+
+                    // Exit the remote session and return results back to RemoteEnumDebugProcess

```

Comment:
As we talked offline, let's enter issue about this. As from customer perspective, we have to avoid this, it is just about how, which leaves it as technical question to us (even it is impossible, we need have more work to get solid prove).